source/agora/cli/checkvtable/main.d

```diff
@@ -0,0 +1,222 @@
+
+            first = true;
+            static foreach (idx, member; __traits(allMembers, %1$s))
+            {
+                mixin(
+                q{
+                    static if (__traits(isVirtualMethod, %1$s.%2$s) && (`%2$s` != `__dtor`) && (`%2$s` != `__xdtor`))
+                    {
+                        writeMethodContents(file, first, `%1$s`, `%2$s`);
+                        first = false;
+                    }
+                }.format(`%1$s`, member));
+            }
+
+            writeMethodFooter(file);
+
+            writeCheckFunction(file, `%1$s`);
+
+            file.writeln(``);
+
+        }.format(e, `%2$s`));
+    }
+
+
+    file.writeln(`int checkVMOffset (const char* classname, const char* offsets)`);
+    file.writeln(`{`);
+
+    static foreach (idx, e; VTableCheckClasses)
+    {
+        mixin(
+        q{
+            if (idx == 0)
+                file.writeln(`    if (strcmp(classname, "%1$s") == 0)`);
+            else
+                file.writeln(`    else if (strcmp(classname, "%1$s") == 0)`);
+            file.writeln(`    {`);
+            file.writeln(`        return checkVMOffset%1$s(offsets);`);
+            file.writeln(`    }`);
+        }.format(e));
+    }
+
+    file.writeln(`    else`);
+    file.writeln(`    {`);
+    file.writeln(`        return -1;`);
+    file.writeln(`    }`);
+    file.writeln(`}`);
+
+    file.close();
+
+    writefln("Source code was generated with the file name %s.", filename);
+
+    return 0;
+}
+
+/// Writes the header of the source file
+void writeHeader(File file)
+{
+    file.writeln(
+`/*******************************************************************************
+
+    Contains checking the offset of virtual methods.
+
+    Note:
+        This was automatically generated by agora-checkvtable.
+        Do not modify this file.
+        If you want to modify it, please use agora-checkvtable.
+        agora-checkvtable source/scpp/extra/DVMChecks.cpp
+        or
+        dub -c checkvtable -- source/scpp/extra/DVMChecks.cpp
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+#include "DVMChecks.h"
+
+#include <iostream>
+#include <string>
+#include <sstream>
+
+using namespace std;
+`
+    );
+}
+
+/// Writes the name of the method
+void writeMethodHeader(File file, string class_name, string method_name)
+{
+    file.writefln(`///  Returns the offset of virtual methods inside the class %s.
+long %s (const char* name)
+{`,
+    class_name, method_name);
+}
+
+/// Writes the end of the method
+void writeMethodFooter(File file)
+{
+    file.writeln(`    else
+    {
+        return -1;
+    }
+}
+`
+    );
+}
+
+/// Writes the contents of the method
+void writeMethodContents(File file, bool first, string entry, string method)
+{
+    if (first)
```

Comment:
You could simplify this by removing the need to use `first`. So instead of generating this:

```C++
if (strcmp(name, "foo") == 0)
{
    auto pf = class::foo;
    return  ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}
else if (strcmp(name, "bar") == 0)
{
    auto pf = class::bar;
    return  ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}
```

You could remove the `first` parameter and generate this:


```C++
if (strcmp(name, "foo") == 0)
{
    auto pf = class::foo;
    return ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}

if (strcmp(name, "bar") == 0)
{
    auto pf = class::bar;
    return ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}
```

The end result is the same (short-circuiting evaluation) as using `if/else if`, because there is a `return` statement in each block.